Add EIP: Revert creation in case of collision (#6733)
ethereum/EIPs: EIPS/eip-creation_collision.md

---

> eip: <not assigned>

Reviewer: ```suggestion
eip: 6733
```

Don't forget to rename the file and address the other `eipw` check annotations as well.

title: Revert creation in case of collision
description: Revert contract creation if address already has code
author: Renan Rodrigues de Souza (@RenanSouza2)
discussions-to: https://ethereum-magicians.org/t/eip-revert-on-address-collision/13442
status: Draft
type: Standards Track
category: Core
created: 2023-03-20
---

## Abstract

> There is no current definition to what happens when a contract creation happens in an address with code already deployed. This fix prevents an attack of deploying a contract code and later changing the code arbitrarily.

Reviewer 2: ```suggestion
There is no current definition to what happens when a contract creation happens in an address with code already deployed. This fix prevents an attack of deploying contract code and later changing the code arbitrarily.
```

Author: Thank you for the remarks, I applied them

## Motivation

In EIP-3607 it was estimated that to create an address collision it would take something about 10 billion USD and this number decreases as computers processing power grow.

## Specification

In contract creation where `new contract address` has a `CODEHASH != EMPTYCODEHASH` MUST have its current environment execution reverted, where `EMPTYCODEHASH = 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470`.

## Rationale

> One of the core tenants of smart contracts is it's code being guranteed not to change, with enougth computing power an attacker can change the code stored in an address to any other code allowing them to remove funds or transfer tokens.

Reviewer 2: ```suggestion
One of the core tenants of smart contracts is its code being guaranteed not to change. With enough computing power an attacker can change the code stored in an address to any other code, allowing them to remove funds or transfer tokens.
```

Reviewer 2: (I was just reading your EIP out of interest and noticed a few tweaks I could suggest.)

Author: Thank you

## Backwards Compatibility

> There is a very small possibility of this collision happening to contract cration by contracts.

Reviewer 3: Typo
```suggestion
There is a very small possibility of this collision happening to contract creation by contracts.
```

Author: Thank you for that, fixed
